src/tools/legacy_station_mapping/legacy_station_mapping.py

```diff
@@ -0,0 +1,246 @@
+            print(f'{key}: {value}')
+    else:
+        print("\nNo multiple mappings found.")
+
+
+# print out stations that couldn't be inferred
+def log_missing(missing_mappings):
+    print("\nMissing Mappings %d:" % len(missing_mappings))
+    pprint.pp(missing_mappings)
+
+
+def format_results(station_id_mapping, missing_station_id_mapping):
+    known_mapping = {}
+    for key, values in station_id_mapping.items():
+        if len(values) > 1:
+            raise RuntimeError("Multiple mappings found. Please resolve.")
+        if len(values) == 0:
+            continue
+        known_mapping[key] = values.pop()
+
+    sorted_known_mappings = dict(
+        sorted(known_mapping.items(), key=lambda i: int(i[0])))
+    results = {
+        'known_mappings': sorted_known_mappings,
+        'missing_mappings': missing_station_id_mapping
+    }
+
+    return results
+
+
+def generate_mapping(filename, max_distance=DEFAULT_MAX_DISTANCE_METERS,
+                     verbose=False, write_to_disk=False):
+    df = load_input_csv(filename)
+    print("Matching stations under %d meters" % max_distance)
+
+    station_id_mapping = {}
+    missing_mappings = {}
+    ride_count = 0
+
+    # Filter the dataframe to ensure the left side of the mapping has an
+    # integer ID less than 1000
+    df['start_id_int'] = pd.to_numeric(df['start_id'], errors='coerce')
+    df_filtered = df[df['start_id_int'].notnull() & (df['start_id_int'] < 1000)]
+
+    # Iterate through the filtered dataframe and create the ID mapping
+    for index, row in tqdm(df_filtered.iterrows(), total=df_filtered.shape[0],
+                           desc="Processing Rows"):
+        start_id = row['start_id']
+        start_lat = row['start_lat']
+        start_lng = row['start_lng']
+        start_station_name = row['start_station_name']
+        ride_count += int(row['count'])
+        closest_id = None
+        closest_distance = float('inf')
+
+        if start_id not in station_id_mapping:
+            station_id_mapping[start_id] = set()
+
+        if start_id in MANUAL_MAPPINGS.keys():
+            station_id_mapping[start_id].add(MANUAL_MAPPINGS[start_id])
+            continue
+
+        for other_index, other_row in df.iterrows():
+            if other_row['start_id'] in df_filtered['start_id'].values:
+                # Ignore rows that are legacy ids
+                continue
+            if index == other_index:
+                # skip self record
+                continue
+
+            other_id = other_row['start_id']
+            other_lat = other_row['start_lat']
+            other_lng = other_row['start_lng']
+            other_station_name = other_row['start_station_name']
+
+            # Short circuit if station names match directly
+            if start_station_name.lower() == other_station_name.lower():
+                station_id_mapping[start_id].add(other_id)
+                closest_distance = float('inf')
+                break
```

Comment:
Couldn't this break too early and skip the rest of the records? I'm worried that there could still be duplicates in the new station ids with different station names (like having '@' vs. 'at'), although that seems to be more of a problem with the old stations (id, name) pairs. If you really trust the new station (id, name) pairs, then this will work great. I might be missing something here though.

Reply:
So the purpose of this tool isn't to necessarily resolve discrepancies between station names and locations. This is strictly to get all old integer based ids (ex:`353`) mapped to an appropriate modern station id `M32016`.

I have a plan for the next step of this pipeline to track and resolve discrepancies between names and coordinates.

Thus far I've only seen a single station name (`Somerville Hospital`) with "bad data" that has a faulty mapping due to this eager matching.

All that said, I'm more than happy to come back to this decision later if it makes more sense for this logic to be more rich.

Reply:
That makes sense! Thanks for the additional context. This PR looks great then.